### PR TITLE
Provide a base class for plugin manager classes.

### PR DESCRIPTION
**Pull request by bangerth**
This is a bit like #1807 in that it provides a base class for the various `Manager` classes. This PR also converts two of the manager class to use the common base class. The remaining ones will follow later.

Observations:
* There isn't quite as much commonality, or at least it's not as easy to refactor.
* A nuisance is that the different `Manager` classes use different names for the same functionality. For example, we have `Postprocessor::Manager<dim>::has_matching_postprocessor()` but `HeatingModels::Manager<dim>::has_matching_heating_model()`. Both do the same, but because they have different names I can't abstract them into the base class (though I have abstracted their *functionality* into the base class).

### CODE REVIEW

**Review by gassmoeller — COMMENTED** [review]
Very nice! I particularly like that the managers are also derived from InterfaceBase. You will have to update one of the tests.

File: include/aspect/plugins.h [hunk 1]
```diff
@@ -188,6 +188,121 @@ namespace aspect
         void
         parse_parameters (ParameterHandler &prm);
     };
+
```
[the remark this answers is not in the document]
Comment (bangerth) [reply include/aspect/plugins.h:191]: Isn't this our usual style for separating top-level declarations (here, two classes)?

File: include/aspect/postprocess/interface.h [hunk 2]
```diff
@@ -345,11 +339,7 @@ namespace aspect
     bool
     Manager<dim>::has_matching_postprocessor () const
     {
-      for (const auto &p : postprocessors)
-        if (Plugins::plugin_type_matches<PostprocessorType>(*p))
-          return true;
-
-      return false;
+      return this->template has_matching_plugin_object<PostprocessorType>();
```
[the remark this answers is not in the document]
Comment (bangerth) [reply include/aspect/postprocess/interface.h:342]: Yes, I think deprecating them is the way to go. They are usually used in context such as `this->get_postprocess_manager().template has_matching_postprocessor<Postprocess::Particles<dim>>()`, in which "postprocess" is used more than once. So I'm not overly concerned about the loss of expressiveness.

**Review by bangerth — COMMENTED** [review]
I addressed as requested. Let's see whether this works.

**Comment by bangerth** [comment]
OK, works now!

File: include/aspect/plugins.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (gassmoeller) [reply include/aspect/plugins.h:191]: Not sure, I think we are not very consistent about top-level separations in header files. I checked `material_model/interface.h` and two lines seems more common, but three lines also exists. I am also fine going with 3 lines (it is more consistent with the source files). Let's not worry about it here.

**Review by gassmoeller — APPROVED** [review]
Nice, another major cleanup :+1:

**Comment by bangerth** [comment]
Well, it's only the beginning. There's substantially more to come to complete this transition.

File: include/aspect/postprocess/interface.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (bangerth) [reply include/aspect/postprocess/interface.h:342]: I'm now tracking this in #1775.

**Pull request merged**